### PR TITLE
`chai`: add type narrowing to `assert.instanceOf` / `assert.notInstanceOf`

### PR DESCRIPTION
**Pull request by arthurschreiber**
This adds type narrowing to `assert.instanceOf` and `assert.notInstanceOf` functions.

I _think_ those type assertions should be correct. 😅

### CODE REVIEW

**Comment by arthurschreiber** [comment]
Mh. The tests for the `assert.notInstanceOf` method are failing because of how the structural typing in TypeScript works. 😞 

Essentially, in the test cases, `Foo` is an empty class. An empty class is structurally "equal" to any other empty class, so when using `Exclude` to remove it from the union type of the `value` argument, other empty classes are removed as well. I don't think there's any way for this to be fixed.

Should I change the tests to use a different setup that does not fail? Or should I remove the type assertion from `assert.notInstanceOf`?

File: types/chai/index.d.ts [hunk 1]
```diff
@@ -782,7 +782,7 @@ declare namespace Chai {
          * @param value    Actual value
          * @param message   Message to display on error.
          */
-        isFinite<T>(value: T, message?: string): void;
+        isFinite<T>(value: T, message?: string): asserts value is (T extends number ? T : never);
```
Comment (amcasey) [inline types/chai/index.d.ts:785]: Isn't `Infinity` typed as `number` in TS?

File: types/chai/index.d.ts [hunk 2]
```diff
@@ -791,7 +791,7 @@ declare namespace Chai {
          * @param value   Actual value.
          * @param message   Message to display on error.
          */
-        isBoolean<T>(value: T, message?: string): void;
+        isBoolean<T>(value: T, message?: string): asserts value is (T extends boolean ? T : never);
```
Comment (amcasey) [inline types/chai/index.d.ts:794]: Is this going to assert `value is never` when `T` is `unknown`?

**Review by amcasey — CHANGES_REQUESTED** [review]

File: types/chai/index.d.ts (hunk 1 above)
Comment (arthurschreiber) [reply types/chai/index.d.ts:785]: But does that matter here? 🤔 `assert.isFinite` throws if the given `value` is either not a number, or is a number but not finite. So after calling `assert.isFinite`, `value` is guaranteed to be a `number`. But I could be misunderstanding your concern? 🙇‍♂️

File: types/chai/index.d.ts (hunk 2 above)
Comment (arthurschreiber) [reply types/chai/index.d.ts:794]: You're right. That's not correct. 🤔 Let me try to improve that.

**Review by arthurschreiber — COMMENTED** [review]
👋 @amcasey 

I added a whole bunch of additional test cases, and fixed up a bunch of issues around the type assertions.

I also had to change a bunch of existing method signatures to simple method overloads instead of using generics, but I think that should be ok, mostly. I don't think some of these methods (like `assert.fail`, `assert.ok`, `assert.notOk` and so on) should've made use of generics in the first place. 🤷 

Would you mind taking another look? 🙇‍♂️

File: types/chai/index.d.ts [hunk 3]
```diff
@@ -791,7 +787,8 @@ declare namespace Chai {
          * @param value   Actual value.
          * @param message   Message to display on error.
          */
-        isBoolean<T>(value: T, message?: string): void;
+        isBoolean(value: object, message?: string): asserts value is never;
+        isBoolean(value: unknown, message?: string): asserts value is boolean;
```
Comment (arthurschreiber) [inline types/chai/index.d.ts:791]: Here's a good example of a method that changed from using generics, to using a simple overload instead. `isBoolean` should never have been generic in the first place. 🤷 

With the overloads, we can make the case where a passed in value is `unknown` (or `any`) work correctly, as the later overload (the one that `asserts value is boolean`) will be used for those types for `value`.

I had to also add the `object` specific overload to handle object-like values correctly. Without that overload asserting that `value is never`, value would end up being asserted as a intersection between the original object type and `boolean`, which doesn't really make sense.

File: types/chai/index.d.ts (hunk 1 above)
Comment (amcasey) [reply types/chai/index.d.ts:781]: Sorry, I was thinking of something else.  You're correct that there's nothing interesting about the `Infinity` case.

File: types/chai/index.d.ts [hunk 4]
```diff
@@ -422,110 +422,101 @@ declare namespace Chai {
...
          *
-         * @type T   Type of the objects.
          * @param actual   Actual value.
          * @param expected   Potential expected value.
          * @param message    Message to display on error.
          * @param operator   Comparison operator, if not strict equality.
          * @remarks Node.js assert module-compatible.
          */
-        fail<T>(actual: T, expected: T, message?: string, operator?: Operator): never;
+        fail(actual: unknown, expected: unknown, message?: string, operator?: Operator): never;
```
Comment (amcasey) [inline types/chai/index.d.ts:431]: The old form would report an error at design-time for `fail("1", 2)`, but the new form doesn't.  Intentional?

Silly question: why would a `fail` function take expected and actual values?

File: types/chai/index.d.ts [hunk 5]
```diff
@@ -422,110 +422,101 @@ declare namespace Chai {
...
         /**
          * Asserts that object is falsy.
          *
-         * @type T   Type of object.
          * @param object   Object to test.
          * @param message    Message to display on error.
          */
-        notOk<T>(value: T, message?: string): void;
+        notOk(value: object, message?: string): asserts value is never;
+        notOk(value: unknown, message?: string): asserts value is "" | 0 | false | null | undefined;
```
Comment (amcasey) [inline types/chai/index.d.ts:465]: Is this a complete list of falsy values?  What about `NaN`?

File: types/chai/index.d.ts [hunk 6]
```diff
@@ -422,110 +422,101 @@ declare namespace Chai {
...
         /**
          * Asserts non-strict equality (==) of actual and expected.
          *
-         * @type T   Type of the objects.
          * @param actual   Actual value.
          * @param expected   Potential expected value.
          * @param message   Message to display on error.
          */
-        equal<T>(actual: T, expected: T, message?: string): void;
+        equal(actual: unknown, expected: unknown, message?: string): void;
```
Comment (amcasey) [inline types/chai/index.d.ts:474]: As above - I think the generic was intended to flag some problems at design time.

File: types/chai/index.d.ts [hunk 7]
```diff
@@ -545,6 +536,8 @@ declare namespace Chai {
          * @param message   Message to display on error.
          */
         isAbove(valueToCheck: number, valueToBeAbove: number, message?: string): void;
+        isAbove(valueToCheck: Date, valueToBeAbove: Date, message?: string): void;
+        isAbove(valueToCheck: unknown, valueToBeAbove: unknown, message?: string): void;
```
Comment (amcasey) [inline types/chai/index.d.ts:540]: What's the `unknown` overload for?  I'm sure the function does something, but does it make sense for anyone to call it that way?

**Review by amcasey — CHANGES_REQUESTED** [review]
I wonder whether intersection types might simplify things a bit.

File: types/chai/index.d.ts [hunk 8]
```diff
@@ -572,24 +569,26 @@ declare namespace Chai {
...
 
         /**
          * Asserts that value is false.
          *
-         * @type T   Type of value.
          * @param value   Actual value.
          * @param message   Message to display on error.
          */
-        isFalse<T>(value: T, message?: string): void;
+        isFalse<T extends object | string | number | boolean | undefined | null>(value: T, message?: string): asserts value is (T extends false ? (false extends T ? false : T) : never);
```
[the remark this answers is not in the document]
Comment (amcasey) [reply types/chai/index.d.ts:590]: Better still, just keep the second overload.

**Comment by arthurschreiber** [comment]
I don't think intersection types make sense to use here (and the simple `asserts value is ...` is doing the exact same thing as an intersection type).

Here's an example using type narrowing with a `typeof` check:

```ts
declare let value: { some: string };

if (typeof value === 'boolean') {
  // never
  value;
}
```

Here's the same example using `assert.isBoolean` using the regular `asserts is boolean` assertion:

```ts
declare let value: { some: string };
declare let assert: {
  isBoolean(value: unknown): asserts value is boolean
}

assert.isBoolean(value);
value; // { some: string } & boolean
```

And here using a generic with intersection:

```ts
declare let value: { some: string };
declare let assert: {
  isBoolean(value: unknown): asserts value is boolean
}

assert.isBoolean(value);
value; // { some: string } & boolean
```

And here using a generic with a type extension check:

```ts
declare let value: { some: string };
declare let assert: {
  isBoolean<T>(value: T): asserts value is (T extends boolean ? T : never)
}

assert.isBoolean(value);
value; // never
```

That last example is the one that mirrors the regular type narrowing behavior of typescript.

---

Now, why did I add the double type extension check? It's specifically for `any`. Without the double check (`T extends boolean` and `boolean extends T`), `any` and doesn't get narrowed correctly:

```ts
declare let value: any;
declare let assert: {
  isBoolean<T>(value: T): asserts value is (T extends boolean ? T : never)
}

assert.isBoolean(value);
value; // any
```

But with the double check we can assert that the type of `value` is `boolean` (instead of just `T`):

```ts
declare let value: any;
declare let assert: {
  isBoolean<T>(value: T): asserts value is (T extends boolean ? (boolean extends T ? boolean : never) : never)
}

assert.isBoolean(value);
value; // boolean
```

---

But there's still one problem here, namely around narrowing `unknown` input values. `unknown` is matched by the generic `T`, passes the `T extends boolean` check but then fails on the `boolean extends T` and ends up being narrowed to `never`.

```ts
declare let value: unknown;
declare let assert: {
  isBoolean<T>(value: T): asserts value is (T extends boolean ? (boolean extends T ? boolean : never) : never)
}

assert.isBoolean(value);
value; // never
```

To fix this, I needed to constrain `T` to all types except `unknown`, and add a second overload specifically for `unknown`.

```ts
declare let value: unknown;
declare let assert: {
  isBoolean<T extends object | string | number | boolean | symbol | bigint | undefined | null>(value: T): asserts value is (T extends boolean ? (boolean extends T ? boolean : never) : never)
  isBoolean(value: unknown): asserts value is boolean
}

assert.isBoolean(value);
value; // boolean
```

**Comment by amcasey** [comment]
Stepping back a bit, what problem is this PR trying to solve?  It's definitely handy to be able to assume something is truthy after a certain point, but things like `isFinite` don't seem to affect the type at all.  And setting the type to `never` won't have any effect since the following code will be unreachable anyway?

**Comment by arthurschreiber** [comment]
I'm maintaining `tedious`, a database driver for Node.JS to connect to Microsoft SQL Server. There's no sensible typing we can do for the things like column values or other results returned from the server side, so a lot of the data returned is typed as `any`.

In the tests, we use `chai` and it would be neat if things like `assert.isString` would make typescript understand that after such a call, the passed value can only be a `string` and nothing else. This is exactly what assertion signatures are for, so it would be nice to have the `chai` type definitions actually make use of this.

The assertion on `assert.isFinite` is maybe a bit less useful, but can still come in handy in general (e.g. if you want to narrow unions between numbers and non-number types). I don't really see an issue with that one in particular, but if it helps I can remove the check for it?

**Review by amcasey — COMMENTED** [review]
I think the positive assertions could be valuable, but most of the negative assertions are difficult or impossible to express accurately in the current type system.  Consider scaling back to the most obviously correct and useful assertions.

If you already have a use case in mind (`tedious`) consider updating your local copy of the .d.ts file to see how it feels in practice.

File: types/chai/index.d.ts [hunk 9]
```diff
@@ -422,110 +422,101 @@ declare namespace Chai {
...
         /**
          * Asserts non-strict inequality (!=) of actual and expected.
          *
-         * @type T   Type of the objects.
          * @param actual   Actual value.
          * @param expected   Potential expected value.
          * @param message   Message to display on error.
          */
-        notEqual<T>(actual: T, expected: T, message?: string): void;
+        notEqual(actual: unknown, expected: unknown, message?: string): void;
```
[the remark this answers is not in the document]
Comment (arthurschreiber) [reply types/chai/index.d.ts:483]: That should be the correct behaviour tho. Again, `assert.equal` / `assert.notEqual` is pretty lax about the types (and should be).

`assert.equal(5, '5')` is valid, and so should be `assert.notEqual(5, '4')`.

File: types/chai/index.d.ts [hunk 10]
```diff
@@ -652,25 +647,23 @@ declare namespace Chai {
          * @param value   Actual value.
          * @param message    Message to display on error.
          */
-        exists<T>(value: T, message?: string): void;
+        exists<T>(value: T, message?: string): asserts value is (T extends undefined | null ? never : T);
```
[the remark this answers is not in the document]
Comment (arthurschreiber) [reply types/chai/index.d.ts:650]: That would assert that the type of `value` is `number`.

File: types/chai/index.d.ts [hunk 11]
```diff
@@ -598,52 +597,48 @@ declare namespace Chai {
          * @param value   Actual value.
          * @param message   Message to display on error.
          */
-        isNotTrue<T>(value: T, message?: string): void;
+        isNotTrue<T>(value: T, message?: string): asserts value is (T extends true ? never : T);
```
[the remark this answers is not in the document]
Comment (arthurschreiber) [reply types/chai/index.d.ts:600]: Yeah, I agree. I'm going to remove the type assertions from the negative assertion methods.

File: types/chai/index.d.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (amcasey) [reply types/chai/index.d.ts:483]: I just like to make sure such changes are intentional. :wink:

File: types/chai/index.d.ts (hunk 4 above)
Comment (amcasey) [reply types/chai/index.d.ts:431]: Doesn't match any of the testing libraries I've used, but sounds intentional.

File: types/chai/index.d.ts (hunk 6 above)
Comment (amcasey) [reply types/chai/index.d.ts:474]: Thanks for confirming!

File: types/chai/index.d.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (amcasey) [reply types/chai/index.d.ts:590]: To confirm, the second overload I was referring to was `isFalse(value: unknown, message?: string): asserts value is false;` and not the intersection.

**Comment by amcasey** [comment]
> I don't think intersection types make sense to use here (and the simple `asserts value is ...` is doing the exact same thing as an intersection type).

I agree.  Sorry for not disavowing that more explicitly.

FWIW, I was expecting `{ foo: 1 } & boolean` to simplify to `never` but it looks like it doesn't (maybe if you assigned something to it?).

File: types/chai/index.d.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (amcasey) [reply types/chai/index.d.ts:650]: Once again, I've been tripped up by the distributive nature of conditional types. :facepalm:

File: types/chai/index.d.ts (hunk 5 above)
Comment (amcasey) [reply types/chai/index.d.ts:465]: I don't think there's a literal type for `NaN` and I'm not sure how else you would express it.

**Pull request closed without merging**